machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,389 @@
+type MachineConfigPoolCondition struct {
+	// Type of the condition, currently ('Done', 'Updating', 'Failed').
+	Type MachineConfigPoolConditionType `json:"type"`
+
+	// Status of the condition, one of ('True', 'False', 'Unknown').
+	Status corev1.ConditionStatus `json:"status"`
+
+	// LastTransitionTime is the timestamp corresponding to the last status
+	// change of this condition.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+
+	// Reason is a brief machine readable explanation for the condition's last
+	// transition.
+	Reason string `json:"reason"`
+
+	// Message is a human readable description of the details of the last
+	// transition, complementing reason.
+	Message string `json:"message"`
+}
+
+// MachineConfigPoolConditionType valid conditions of a machineconfigpool
+type MachineConfigPoolConditionType string
+
+const (
+	// MachineConfigPoolUpdated means machineconfigpool is updated completely.
+	// When the all the machines in the pool are updated to the correct machine config.
+	MachineConfigPoolUpdated MachineConfigPoolConditionType = "Updated"
+	// MachineConfigPoolUpdating means machineconfigpool is updating.
+	// When at least one of machine is not either not updated or is in the process of updating
+	// to the desired machine config.
+	MachineConfigPoolUpdating MachineConfigPoolConditionType = "Updating"
+	// MachineConfigPoolNodeDegraded means the update for one of the machine is not progressing
+	MachineConfigPoolNodeDegraded MachineConfigPoolConditionType = "NodeDegraded"
+	// MachineConfigPoolRenderDegraded means the rendered configuration for the pool cannot be generated because of an error
+	MachineConfigPoolRenderDegraded MachineConfigPoolConditionType = "RenderDegraded"
+	// MachineConfigPoolDegraded is the overall status of the pool based, today, on whether we fail with NodeDegraded or RenderDegraded
+	MachineConfigPoolDegraded MachineConfigPoolConditionType = "Degraded"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigPoolList is a list of MachineConfigPool resources
+type MachineConfigPoolList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineConfigPool `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KubeletConfig describes a customized Kubelet configuration.
+type KubeletConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   KubeletConfigSpec   `json:"spec,omitempty"`
+	Status KubeletConfigStatus `json:"status,omitempty"`
+}
+
+// KubeletConfigSpec defines the desired state of KubeletConfig
+type KubeletConfigSpec struct {
+	MachineConfigPoolSelector *metav1.LabelSelector                      `json:"machineConfigPoolSelector,omitempty"`
+	KubeletConfig             *kubeletconfigv1beta1.KubeletConfiguration `json:"kubeletConfig,omitempty"`
```

Comment:
would a RawExtension make more sense? Makes you independent from the actual kubelet config version.

Reply:
cc @rphillips (but I believe we can postpone as well)

Reply:
You're exposing this as part of v1 API, postponing means you won't get a chance to do it, ever.